### ISSUE TITLE
DataSchemaValue in TS should not be any

### ISSUE THREAD

**Issue by relu91** [w3c/wot-scripting-api#296]
As noted in this comment https://github.com/w3c/wot-scripting-api/issues/294#issuecomment-766912927 DataSchemaValue in the TS definitions is an alias type for any. I think we should improve the definition since `any` type allows any kind of assignment (i.e., a function or a Symbol). I think we should restrict it to the valid JSONSchema types: `string, number, array, object, null`.

**Comment by zolkis** [comment]
In Web IDL it's simpler to use `any` and specify validation ([JSON Schema types](https://json-schema.org/understanding-json-schema/reference/type.html), including `boolean`) in the algorithms. Otherwise we'd need to make a union type of like 18 [Web IDL types](https://heycam.github.io/webidl/#idl-types).

**Comment by danielpeintner** [comment]
Another alternative would be to have the union type for Typescript only... not sure if this divergence is worth it though...

`type DataSchemaValue = (string | number  | array | object | null );`

**Comment by relu91** [comment]
> Another alternative would be to have the union type for Typescript only... not sure if this divergence is worth it though...
> 
> `type DataSchemaValue = (string | number | array | object | null );`

That's exactly what I was thinking about. Not sure if it worths translating it in WebIDL since there as @zolkis said we could use an algorithm to enforce a particular type.

**Comment by danielpeintner** [comment]
> That's exactly what I was thinking about. Not sure if it worths translating it in WebIDL since there, as @zolkis said, we could use an algorithm to enforce a particular type constrain.

We state [here](https://w3c.github.io/wot-scripting-api/#the-interactioninput-type) in prose the possible types.

Making the TypeScript definitions more accurate as described makes perfect sense to me!

**Comment by relu91** [comment]
Just for the record, the WebIDL would look something like this:
```idl
typedef (object or double or boolean or USVString or FrozenArray<DataSchemaValue> or undefined)? DataSchemaValue;
```

But we still have to use some text to explain the mapping between WebIDL types and JSONSchema types (i.e., double -> number, USVString -> string).

**Comment by zolkis** [comment]
No, `double` is not enough, we have a lot more types in Web IDL. The guidance is to use `any` in Web IDL and explain the usage in prose.

**Comment by danielpeintner** [comment]
It seems best to me to improve TS definitions while we might need to live with the current WebIDL (and the types described in prose).

Is it OK for everyone if I go ahead and provide a PR for the TD definitions?

**Comment by relu91** [comment]
> Is it OK for everyone if I go ahead and provide a PR for the TD definitions?

Ok for me!

> No, double is not enough, we have a lot more types in Web IDL. The guidance is to use any in Web IDL and explain the usage in prose.

I know but most of them are not valid JSONSchema types. `double` would work fine for JSONSchema `number`.

**Comment by zolkis** [comment]
For WoT Scripting (non-browser implementation) the Web IDL in not crucial. FWIW if we need to mape the generic `number`, we could use `unrestricted double` in Web IDL. But I prefer using `any` and specify in the algorithm when is it boolean, number, string, object, array, null, NaN, undefined, etc.

**Comment by danielpeintner** [comment]
> > Is it OK for everyone if I go ahead and provide a PR for the TD definitions?
> 
> Ok for me!

I created PR https://github.com/w3c/wot-scripting-api/pull/297 but I also raised a question in the pull request.

**Comment by relu91** [comment]
I think we can close this issue. It explicitly refers to TS types and the change was addressed by #297. About WebIDL we agreed to leave the exact description of the DataSchemaValue type to prose.
